### PR TITLE
Add: duckdb vector store

### PR DESCRIPTION
**Pull request by krish-adi**
# Description

This PR adds DuckDB as a vector store. 

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+requires = ["poetry-core"]
+build-backend = "poetry.core.masonry.api"
+
+[tool.codespell]
+check-filenames = true
+check-hidden = true
+# Feel free to un-skip examples, and experimental, you will just need to
+# work through many typos (--write-changes and --interactive will help)
+skip = "*.csv,*.html,*.json,*.jsonl,*.pdf,*.txt,*.ipynb"
+
```
Comment (nerdai) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/pyproject.toml:11]: Hey I know this wasn't yet specified in our CONTRIBUTING guidelines, but nevertheless can you please add a section for `[tool.llamahub]` the one found [here](https://github.com/run-llama/llama_index/blob/8431e7baaeb8155c33a7db328fce79bd7d533feb/llama-index-integrations/vector_stores/llama-index-vector-stores-astra/pyproject.toml#L10)

File: llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+# Remove venv skip when integrated with pre-commit
+exclude = ["_static", "build", "examples", "notebooks", "venv"]
+ignore_missing_imports = true
+python_version = "3.8"
+
+[tool.poetry]
+name = "llama-index-vector-stores-duckdb"
+version = "0.1.0"
+description = "llama-index vector_stores duckdb integration"
+authors = ["Adithya Krishnan <me@krishadi.com>"]
```
Comment (nerdai) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/pyproject.toml:23]: Can you also please add yourself as a maintainer like [this](https://github.com/run-llama/llama_index/blob/8431e7baaeb8155c33a7db328fce79bd7d533feb/llama-index-packs/llama-index-packs-agents-llm-compiler/pyproject.toml#L26). Please use your github username.

File: llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+license = "MIT"
+readme = "README.md"
+packages = [{include = "llama_index/"}]
+
+[tool.poetry.dependencies]
+python = ">=3.8.1,<3.12"
```
Comment (nerdai) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/pyproject.toml:29]: We haven't yet fixed the `new-package` tool, but please change this `">=3.8.1<4.0"`

File: llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/tests/test_vector_stores_duckdb.py
Comment (nerdai) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/tests/test_vector_stores_duckdb.py:1]: Can you kindly add a test similar to this one: https://github.com/run-llama/llama_index/tree/main/llama-index-integrations/vector_stores/llama-index-vector-stores-astra/tests

**Review by nerdai — COMMENTED** [review]
Thanks @krish-adi! Looks pretty good. Left a few minor comments

File: llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/pyproject.toml (hunk 1 above)
Comment (krish-adi) [reply llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/pyproject.toml:11]: fixed this.

File: llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/pyproject.toml (hunk 2 above)
Comment (krish-adi) [reply llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/pyproject.toml:25]: fixed this.

File: llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/tests/test_vector_stores_duckdb.py
Comment (krish-adi) [reply llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/tests/test_vector_stores_duckdb.py:1]: fixed this.

File: llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/pyproject.toml (hunk 3 above)
Comment (krish-adi) [reply llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/pyproject.toml]: fixed this.

**Comment by krish-adi** [comment]
@nerdai made updates and committed the changes with the tests.

**Review by nerdai — APPROVED** [review]
LGTM!

**Comment by nerdai** [comment]
Looks like gotta run pants to tailor some BUILD files. Am AFK at moment -- will do so in a few hours.

**Comment by nerdai** [comment]
@krish-adi can you add me to your fork/branch? I ran pants on my end and would like to push the new files.

**Comment by krish-adi** [comment]
@nerdai cool, added you as a collaborator to my fork...

**Comment by nerdai** [comment]
@krish-adi, I was able to get past the pants issue for not running the tests, but now its failing due another error. See traceback attached below. When I run the test locally, everything passes. The error is indicating something about a home directory not being found — perhaps due to our runner configuration and our pants ci/cd setup. Just checking with you if know there is a way to set up a home directory for `DuckDBVectorStore`?



```python
    @pytest.fixture(scope="module")
    def vector_store() -> DuckDBVectorStore:
>       return DuckDBVectorStore()

llama-index-integrations/vector_stores/llama-index-vector-stores-duckdb/tests/test_duckdb.py:85: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = DuckDBVectorStore(), database_name = ':memory:', table_name = 'documents'
embed_dim = 1536
text_search_config = {'ignore': '(\\.|[^a-z])+', 'lower': True, 'overwrite': False, 'stemmer': 'english', ...}
persist_dir = './storage', kwargs = {}
duckdb = <module 'duckdb' from '/home/runner/.cache/pants/named_caches/pex_root/venvs/s/181b35f8/venv/lib/python3.8/site-packages/duckdb/__init__.py'>

    def __init__(
        self,
        database_name: Optional[str] = ":memory:",
        table_name: Optional[str] = "documents",
        # schema_name: Optional[str] = "main",
        embed_dim: Optional[int] = 1536,
        # hybrid_search: Optional[bool] = False,
        # https://duckdb.org/docs/extensions/full_text_search
        text_search_config: Optional[dict] = {
            "stemmer": "english",
            "stopwords": "english",
            "ignore": "(\\.|[^a-z])+",
            "strip_accents": True,
            "lower": True,
            "overwrite": False,
        },
        persist_dir: Optional[str] = "./storage",
        **kwargs: Any,
    ) -> None:
        """Init params."""
        try:
            import duckdb
        except ImportError:
            raise ImportError(import_err_msg)
    
        self._is_initialized = False
    
        if database_name == ":memory:":
            self._conn = duckdb.connect(database_name)
>           self._conn.install_extension("json")
E           duckdb.duckdb.IOException: IO Error: Can't find the home directory at ''
E           Specify a home directory using the SET home_directory='/path/to/dir' option.****
```

**Comment by krish-adi** [comment]
@nerdai just commited a change to set the home_directory. This uses assumes that the home directory is the same as the one the python interpretor has access to as the home dir.

**Comment by nerdai** [comment]
> @nerdai just commited a change to set the home_directory. This uses assumes that the home directory is the same as the one the python interpretor has access to as the home dir.

@krish-adi that did it -- nice job! Thanks for the contribution!!! Merging now.

**Comment by nerdai** [comment]
@krish-adi I actually just noticed that you already went ahead an published the package. No problem. But could you also add ` jerryjliu` as a collaborator (with the `owner` role) to the pypi project? In case you have not done this before, you can do that by logging into your PyPi account and clicking manage on the `llama-index-vector-stores-duckdb` package. Then on the left sidebar click Collaborators and from there you should be able to add `jerryjliu` to the project 🙏.

**Comment by krish-adi** [comment]
@nerdai added jerryliu to the pypi project!

**Pull request merged**